src/bsoncxx/string/view_or_value.hpp

```diff
@@ -0,0 +1,83 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/config/prelude.hpp>
+
+#include <string>
+
+#include <bsoncxx/stdx/string_view.hpp>
+#include <bsoncxx/view_or_value.hpp>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+namespace string {
+
+///
+/// Class representing a view-or-value variant type for strings.
+///
+/// This class adds several string-specific methods to the bsoncxx::view_or_value template:
+/// - a constructor overload for const char*
+/// - a constructor overload for std::string by l-value reference, for convenience
+/// - a safe c_str() operation to return null-terminated c-style strings.
+///
+class BSONCXX_API view_or_value : public bsoncxx::view_or_value<stdx::string_view, std::string> {
+   public:
+    ///
+    /// Forward all bsoncxx::view_or_value constructors.
+    ///
+    using bsoncxx::view_or_value<stdx::string_view, std::string>::view_or_value;
+
+    ///
+    /// Construct a string::view_or_value using a null-terminated const char *.
+    ///
+    /// @param str A null-terminated string
+    ///
+    BSONCXX_INLINE view_or_value(const char* str)
+        : bsoncxx::view_or_value<stdx::string_view, std::string>(std::string(str)) {
+    }
+
+    ///
+    /// Allow construction with an l-value reference to a std::string. The resulting
+    /// view_or_value will keep a string_view of the std::string, so it is important
+    /// that the passed-in string outlive this object.  This constructor is intended
+    /// as a convenience for users.
+    ///
+    /// Construction calls passing a std::string by r-value reference will use the
+    /// constructor defined in the parent view_or_value class.
+    ///
+    /// @param str A std::string l-value reference.
+    ///
+    BSONCXX_INLINE view_or_value(const std::string& str)
+        : bsoncxx::view_or_value<stdx::string_view, std::string>(stdx::string_view(str)) {
+    }
+
+    ///
+    /// Return a null-terminated string from this string::view_or_value.
+    ///
+    /// @note if we are non-owning, we cannot assume our string_view is null-terminated.
+    /// In this case we must make a new, owning view_or_value with a null-terminated
+    /// copy of our view.
+    ///
+    /// @return A pointer to the original string, or a null-terminated copy.
+    ///
+    const char* c_str();
+};
+
```

Comment:
So, I took a look at the base view_or_value template again. I think we are missing some equality operations. In particular, I think we definitely want to add:

```
template <typename View, typename Value>
BSONCXX_INLINE bool operator==(const view_or_value<View, Value>& lhs, const view_or_value<View, Value>& rhs)
```
and the associated `!=`.

However, given that we require that the `View` type `is_constructible` from `Value`, I think it might also be good to add the various permutations around:

```
template <typename View, typename Value>
BSONCXX_INLINE bool operator==(const view_or_value<View, Value>& lhs, const Value& rhs)
```
and the associated rhs/lhs swaps and the `!=` variants.

I still think you want to manually add the (in)equality overloads for the string specialization of view_or_value against `const char*`, given the special relationship between `const char*` and `std::string`